core/lens/lens_phrase.ml

```diff
@@ -0,0 +1,235 @@
+open Utility
+open Types
+open Sugartypes
+open Lens_operators
+open Lens_utility
+
+module Alias = Lens_alias
+
+type t = lens_phrase
+
+let var n = Var n
+
+let and' v1 v2 = InfixAppl (Binary.Logical Logical_binop.And, v1, v2)
+
+let or' v1 v2 = InfixAppl (Binary.Logical Logical_binop.Or, v1, v2)
+
+let equal v1 v2 = InfixAppl (Binary.Equal, v1, v2)
+
+let not' v1 = UnaryAppl (Unary.Not, v1)
+
+let tuple v = TupleLit v
+
+let tuple_singleton v = tuple [v]
+
+let name_of_var expr =
+  match expr.node with
+  | `Var n -> n
+  | _ -> failwith "Expected var."
+
+let of_phrase p =
+  let rec f p =
+    match p.node with
+    | `Constant c -> Constant c
+    | `Var v -> Var v
+    | `UnaryAppl ((_, op), phrase) -> UnaryAppl (Unary.from_links op, f phrase)
+    | `InfixAppl ((_, op), phrase1, phrase2) -> InfixAppl (Binary.from_links op, f phrase1, f phrase2)
+    | `TupleLit l -> TupleLit (List.map f l)
+    | `FnAppl (fn, arg) ->
+      begin
+        match name_of_var fn with
+        | "not" -> UnaryAppl ((Unary.Name "!"), f (List.hd arg))
+        | _ -> failwith "Unsupported function"
+      end
+    | _ -> failwith "Unknown phrasenode for lens_phrase to phrase."
+  in
+  f p
+
+let rec traverse (expr : lens_phrase) dosth =
+  let fn expr' = traverse expr' dosth in
+  let expr = match expr with
+    | Constant _ -> expr
+    | Var _ -> expr
+    | UnaryAppl (a, arg) ->
+      let arg = fn arg in
+      UnaryAppl (a, arg)
+    | InfixAppl (a, a1, a2) ->
+      let a1 = fn a1 in
+      let a2 = fn a2 in
+      InfixAppl (a, a1, a2)
+    | TupleLit (x :: []) ->
+      let x = fn x in
+      TupleLit ([x])
+    | Case (phr, cases, otherwise) ->
+      let phr = OptionUtils.opt_map fn phr in
+      let cases = List.map (fun (inp, lst) -> fn inp, fn lst) cases in
+      let otherwise = fn otherwise in
+      Case (phr, cases, otherwise)
+    | _ -> failwith "Unknown operation" in
+  dosth expr
+
+let get_vars expr =
+  let cols = ref Alias.Set.empty in
+  let _ = traverse expr (fun expr ->
+      match expr with
+      | Var n -> cols := Alias.Set.add n (!cols); expr
+      | _ -> expr) in
+  !cols
+
+let rename_var expr ~replace =
+  traverse expr (fun expr ->
+      match expr with
+      | Var key ->
+        Alias.Map.find ~key replace
+        |> Option.map ~f:var
+        |> Option.value ~default:expr
+      | _ -> expr
+    )
+
+module Constant = struct
+  open Lens_constant
+
+  let bool v = Constant (bool v)
+
+  let int v = Constant (int v)
+
+  let of_value v = Constant (of_value v)
+end
+
+let replace_var expr ~replace =
+  traverse expr (fun expr ->
+      match expr with
+      | Var key ->
+        Alias.Map.find ~key replace
+        |> Option.map ~f:Constant.of_value
+        |> Option.value ~default:expr
+      | _ -> expr
+    )
+
+let rec eval expr get_val =
+  let open Value in
+  match expr with
+  | Constant c -> Lens_constant.to_value c
+  | Var v ->
+    begin
+      try
```

Comment:
Nitpick: You seem to want to save lines below. Here is an opportunity to save some:
```ocaml
begin try get_val v 
      with Not_found -> failwith ("Could not find coiumn " ^ v ^ ".") end
```

Reply:
changed.